File: ios/TNGuardTricare/Models/ContentModel.swift
import Foundation

/// Mirrors content/content.schema.json. Keep field names in sync with content.json.
struct AppContent: Codable, Equatable {
    var contentVersion: Int
    var updatedAt: String
    var trs: TRSContent
    var tnReimbursement: ReimbursementContent
    var faq: [FAQItem]
    var disclaimer: Disclaimer

    struct TRSContent: Codable, Equatable {
        var premiums: Premiums
        var links: TRSLinks
        var phoneNumbers: PhoneNumbers
        var eligibility: [String]
        var steps: [Step]

        struct Premiums: Codable, Equatable {
            var memberOnlyMonthly: String
            var memberAndFamilyMonthly: String
            var asOfYear: Int
        }

        struct TRSLinks: Codable, Equatable {
            var milconnect: String
            var bwe: String
            var trsOverview: String
            var trsEnrollmentSteps: String
            var formsPage: String
            var whenCoverageBegins: String
            var costs: String
            var handbook: String
            var eastRegion: String
            var humanaMilitary: String
            var findDoctor: String

            subscript(key: String) -> String? {
                switch key {
                case "milconnect": return milconnect
                case "bwe": return bwe
                case "trsOverview": return trsOverview
                case "trsEnrollmentSteps": return trsEnrollmentSteps
                case "formsPage": return formsPage
                case "whenCoverageBegins": return whenCoverageBegins
                case "costs": return costs
                case "handbook": return handbook
                case "eastRegion": return eastRegion
                case "humanaMilitary": return humanaMilitary
                case "findDoctor": return findDoctor
                default: return nil
                }
            }
        }

        struct PhoneNumbers: Codable, Equatable {
            var humanaEastLabel: String
            var humanaEast: String
            var triWestLabel: String
            var triWest: String
        }
    }

    struct ReimbursementContent: Codable, Equatable {
        var email: String
        var monthlyDeadlineRule: String
        var links: ReimbursementLinks
        var eligibility: [String]
        var steps: [Step]
        var forms: [FormDefinition]

        struct ReimbursementLinks: Codable, Equatable {
            var programPage: String
            var policyPdf: String
            var enrollmentPacketPdf: String
            var w4Pdf: String
        }
    }

    struct Step: Codable, Equatable, Identifiable {
        var id: String
        var title: String
        var body: String
        var actionLabel: String?
        var actionLinkKey: String?
    }

    struct FormDefinition: Codable, Hashable, Identifiable {
        var id: String
        var title: String
        var description: String
        var pdfTemplate: String
        var isPlaceholderTemplate: Bool?
        /// "acroform" (real interactive PDF fields, matched by name) or "overlay" (a real but
        /// non-interactive/flat page — values are drawn as text on top of it). Defaults to
        /// "acroform" when absent so older content stays valid.
        var fillMode: String?
        /// 0-based page to extract into the final shared PDF. When nil, the whole bundled
        /// document is kept (e.g. the W-4, whose extra pages are legitimate IRS instructions).
        var outputPage: Int?
        var fields: [FormField]

        var resolvedFillMode: FillMode { fillMode == "overlay" ? .overlay : .acroform }

        enum FillMode { case acroform, overlay }
    }

    struct FormField: Codable, Hashable, Identifiable {
        var id: String
        var label: String
        var type: FieldType
        /// Required for acroform fields — the real PDF field name to set.
        var pdfFieldName: String?
        /// Drives a Picker in the UI. For acroform fields these must exactly match the PDF's
        /// export values (e.g. the real dropdown options on the enrollment form).
        var options: [String]?
        var sensitive: Bool?
        var placeholder: String?
        /// Overlay mode only: where to stamp this field's value (or the same value in more
        /// than one spot, e.g. a name repeated in a sentence and in a signature block).
        var overlays: [OverlayPosition]?
        /// Overlay mode only, for choice-type fields: where to draw a mark for each option.
        var overlayOptions: [OverlayChoice]?

        enum FieldType: String, Codable, Hashable {
            case text, number, date, ssn, bankRouting, bankAccount, checkbox, signatureName, choice
        }
    }

    struct OverlayPosition: Codable, Hashable {
        var page: Int
        var x: Double
        var y: Double
    }

    struct OverlayChoice: Codable, Hashable {
        var value: String
        var page: Int
        var x: Double
        var y: Double
    }

    struct FAQItem: Codable, Equatable, Identifiable {
        var id: String { question }
        var question: String
        var answer: String
    }

    struct Disclaimer: Codable, Equatable {
        var title: String
        var body: String
    }
}

File: android/app/src/main/java/org/tnguardtricare/app/model/ContentModel.kt
package org.tnguardtricare.app.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Mirrors content/content.schema.json — keep field names in sync with content.json and with
 * the iOS ContentModel.swift, since both apps read the same file.
 */
@Serializable
data class AppContent(
    val contentVersion: Int,
    val updatedAt: String,
    val trs: TrsContent,
    val tnReimbursement: ReimbursementContent,
    val faq: List<FaqItem>,
    val disclaimer: Disclaimer,
) {
    @Serializable
    data class TrsContent(
        val premiums: Premiums,
        val links: TrsLinks,
        val phoneNumbers: PhoneNumbers,
        val eligibility: List<String>,
        val steps: List<Step>,
    ) {
        @Serializable
        data class Premiums(
            val memberOnlyMonthly: String,
            val memberAndFamilyMonthly: String,
            val asOfYear: Int,
        )

        @Serializable
        data class TrsLinks(
            val milconnect: String,
            val bwe: String,
            val trsOverview: String,
            val trsEnrollmentSteps: String,
            val formsPage: String,
            val whenCoverageBegins: String,
            val costs: String,
            val handbook: String,
            val eastRegion: String,
            val humanaMilitary: String,
            val findDoctor: String,
        ) {
            fun byKey(key: String): String? = when (key) {
                "milconnect" -> milconnect
                "bwe" -> bwe
                "trsOverview" -> trsOverview
                "trsEnrollmentSteps" -> trsEnrollmentSteps
                "formsPage" -> formsPage
                "whenCoverageBegins" -> whenCoverageBegins
                "costs" -> costs
                "handbook" -> handbook
                "eastRegion" -> eastRegion
                "humanaMilitary" -> humanaMilitary
                "findDoctor" -> findDoctor
                else -> null
            }
        }

        @Serializable
        data class PhoneNumbers(
            val humanaEastLabel: String,
            val humanaEast: String,
            val triWestLabel: String,
            val triWest: String,
        )
    }

    @Serializable
    data class ReimbursementContent(
        val email: String,
        val monthlyDeadlineRule: String,
        val links: ReimbursementLinks,
        val eligibility: List<String>,
        val steps: List<Step>,
        val forms: List<FormDefinition>,
    ) {
        @Serializable
        data class ReimbursementLinks(
            val programPage: String,
            val policyPdf: String,
            val enrollmentPacketPdf: String,
            val w4Pdf: String,
        )
    }

    @Serializable
    data class Step(
        val id: String,
        val title: String,
        val body: String,
        val actionLabel: String? = null,
        val actionLinkKey: String? = null,
    )

    @Serializable
    data class FormDefinition(
        val id: String,
        val title: String,
        val description: String,
        val pdfTemplate: String,
        val isPlaceholderTemplate: Boolean? = null,
        /** "acroform" (default) or "overlay" — see FillMode. */
        val fillMode: String? = null,
        /** 0-based page to extract into the final shared PDF; null keeps the whole document. */
        val outputPage: Int? = null,
        val fields: List<FormField>,
    ) {
        val resolvedFillMode: FillMode
            get() = if (fillMode == "overlay") FillMode.OVERLAY else FillMode.ACROFORM
    }

    enum class FillMode { ACROFORM, OVERLAY }

    @Serializable
    data class FormField(
        val id: String,
        val label: String,
        val type: FieldType,
        /** Required for acroform fields — the real PDF field name to set. */
        val pdfFieldName: String? = null,
        /** Drives a dropdown in the UI; for acroform fields must match the PDF's export values. */
        val options: List<String>? = null,
        val sensitive: Boolean? = null,
        val placeholder: String? = null,
        /** Overlay mode only: where to stamp this field's value (possibly in multiple spots). */
        val overlays: List<OverlayPosition>? = null,
        /** Overlay mode only, for choice fields: where to draw a mark for each option. */
        val overlayOptions: List<OverlayChoice>? = null,
    )

    @Serializable
    enum class FieldType {
        @SerialName("text") TEXT,
        @SerialName("number") NUMBER,
        @SerialName("date") DATE,
        @SerialName("ssn") SSN,
        @SerialName("bankRouting") BANK_ROUTING,
        @SerialName("bankAccount") BANK_ACCOUNT,
        @SerialName("checkbox") CHECKBOX,
        @SerialName("signatureName") SIGNATURE_NAME,
        @SerialName("choice") CHOICE,
    }

    @Serializable
    data class OverlayPosition(
        val page: Int,
        val x: Double,
        val y: Double,
    )

    @Serializable
    data class OverlayChoice(
        val value: String,
        val page: Int,
        val x: Double,
        val y: Double,
    )

    @Serializable
    data class FaqItem(
        val question: String,
        val answer: String,
    )

    @Serializable
    data class Disclaimer(
        val title: String,
        val body: String,
    )
}
